### ISSUE TITLE
openvpn 2.4.4 crashes on Mac OS X 10.5.8

### ISSUE THREAD

**Issue by sideeffect42** [Tunnelblick/Tunnelblick#422]
The `openvpn` 2.4.4 binary shipped with Tunnelblick crashes when Tunnelblick is launched on Leopard (probably also on Tiger).

```
Process:         openvpn [11349]
Path:            /Applications/Tunnelblick.app/Contents/Resources/openvpn/openvpn-2.4.4/openvpn
Identifier:      openvpn
Version:         ??? (???)
Code Type:       PPC (Native)
Parent Process:  Tunnelblick [11342]

Date/Time:       2017-12-26 15:31:24.493 +0000
OS Version:      Mac OS X 10.5.8 (9L31a)
Report Version:  6
Anonymous UUID:  261EA74E-97A5-4C5A-9B64-A8418B758076

Exception Type:  EXC_BREAKPOINT (SIGTRAP)
Exception Codes: 0x0000000000000001, 0x000000008fe0105c
Crashed Thread:  0

Dyld Error Message:
  Library not loaded: /usr/lib/libssl.0.9.8.dylib
  Referenced from: /Applications/Tunnelblick.app/Contents/Resources/openvpn/openvpn-2.4.4/openvpn
  Reason: image not found
```

Leopard ships with OpenSSL 0.9.7 preinstalled, and not 0.9.8

```
$ /usr/bin/openssl version
OpenSSL 0.9.7l 28 Sep 2006
$ find /usr/lib /usr/local/lib -name 'libssl*'
/usr/lib/libssl.0.9.7.dylib
/usr/lib/libssl.0.9.dylib
/usr/lib/libssl.dylib
```

Tunnelblick should either ship its own version of OpenSSL or allow to select an external version of OpenVPN (e.g. installed through MacPorts)

**Workaround:** Delete OpenVPN 2.4.x from Tunnelblick and revert the default back to 2.3.x
```
$ cd /Applications/Tunnelblick.app/Contents/Resources/openvpn/
$ sudo rm -rv openvpn-2.4.4
$ sudo rm openvpn
$ sudo ln -s openvpn-2.3.18/openvpn openvpn
```

**Comment by jkbullard** [comment]
@sideeffect42 - Several versions of Tunnelblick ship with OpenVPN 2.4.4. What version of Tunnelblick is crashing?

**Comment by sideeffect42** [comment]
Sorry, I forgot to mention the Tunnelblick version. I'm running Tunnelblick 3.5.19 (build 4270.4970).

Tunnelblick >= 3.5.13 are also affected.

**Comment by jkbullard** [comment]
@sideeffect42 - Thanks, but I'm puzzled: Why change the Tunnelblick binary?
 1.  Changing the binary invalidates the digital signature.
 2. The "default" link already points to 2.3.18. (These old versions of Tunnelblick do not use the 'default" link at all. It is there for use by third-party installers and scripts.)
 
Usually 2.3.18 is also the default. The only way 2.4.4 can become the default is If you had set Tunnelblick to use a specific version of OpenVPN and that version isn't available, Tunnelblick will use the latest version of OpenVPN (2.4.4 in this case), and it will pop up a window telling you about that.

It is better select all of your configurations in the left side of the "Configurations" panel of Tunnelblick's "VPN Details" window, and then set the version of OpenVPN that you want to "Default (OpenVPN 2.3.18)" on the "Settings" tab. That way Tunnelblick will default to use OpenVPN 2.3.18, and if some future version of Tunnelblick updates to have, say, 2.3.19 instead of  2.3.18, it will default to that.

**Comment by sideeffect42** [comment]
I don't know what you mean by "changing the Tunnelblick binary".
Do you mean the workaround I mentioned? That's just a quick hack I figured solves the problem for me (I thought it could be useful for someone else having a similar problem).
Yes, this would invalidate the digital signature, but the signature is broken anyway (but that's a different issue).

I just checked my configurations. They're all set to "Default (OpenVPN 2.3.18)".
I don't know why Tunnelblick even wants to start the openvpn binary when it launches.

The real issue here is that Tunnelblick ships an openvpn binary that is linked against a library that is not present on the system, therefore I suggested shipping a newer version of OpenSSL as well.
One might also consider statically linking openvpn.

**Comment by jkbullard** [comment]
@sideeffect42 - I'm sorry for my earlier confusion. Thanks for your workaround.

This bug crashes Tunnelblick at launch on pre-10.6.8 systems. It went unreported for seven months, so I conclude that almost nobody is using new versions of Tunnelblick on such old systems.

Rather than putting in the time and effort required to fix the problem, I am removing OpenVPN 2.4 support on such old systems. (I'd be happy to consider a PR that would fix it, of course.) Tunnelblick will continue to include the 2.4 OpenVPN binary, but on systems earlier than 10.6.8 it will be ignored and be unavailable.

Commit 222e1754cb75e does this by ignoring non-2.3 versions of OpenVPN on versions of OS X earlier than 10.6.8.

I have updated Tunnelblick's [Deprecated Downloads page](https://tunnelblick.net/downloadsDeprecated.html) to reflect this problem; the page will be updated again when the next Tunnelblick 3.5 version is released with this fix.

> I don't know why Tunnelblick even wants to start the openvpn binary when it launches.

At launch, Tunnelblick checks that each copy of OpenVPN that it includes is named with the version number. This check was added years ago to detect problems when users added their own versions of OpenVPN to Tunnelblick (before digital signatures made that less useful). To do the check, Tunnelblick invokes OpenVPN with the "--version" option and examines the output.

> The real issue here is that Tunnelblick ships an openvpn binary that is linked against a library that is not present on the system, therefore I suggested shipping a newer version of OpenSSL as well.
>One might also consider statically linking openvpn.

Tunnelblick does statically link to SSL libraries. There is a bug that prevents that for the copy OpenVPN 2.4 in Tunnelblick 3.5.*.

**Comment by sideeffect42** [comment]
Thank you so much for still bothering with old versions of OS X. This can't be taken for granted these days.

I don't use VPN on my Mac that often, so I think disabling OpenVPN 2.4 is a sensible solution for the moment.
I'll have a look at it when I have some time and see if I can get it to work.

One reason why it went undiscovered for so long could be that the updater seems to be broken for the legacy version. 
The Sparkle feed at https://www.tunnelblick.net/appcast-s.rss only returns the latest 3.7.x version.

I've been on 3.5.11 (iirc) until some days ago when I checked for a newer version manually.
It could be that there are actually more users of Tunnelblick on old versions of Mac OS X, but haven't updated.

**Comment by jkbullard** [comment]
I have released Tunnelblick 3.5.20, which fixes the problem by making non-2.3 versions of OpenVPN unavailable on OS X versions earlier than 10.6.8.

> One reason why it went undiscovered for so long could be that the updater seems to be broken for the legacy version.

You're correct, updating Tunnelblick 3.5 is broken. Why it's broken is a long story, but I have an idea that might work to fix it.

**If you are interested in keeping old (3.5) versions of Tunnelblick up-to-date** in the meantime, you should subscribe to the [Tunnelblick Announce mailing list](https://groups.google.com/forum/#!forum/tunnelblick-announce). All releases are announced on that low volume list (18 announcements in 2017, for example).

**Comment by sideeffect42** [comment]
Thanks for fixing all of the issues I mentioned. Very nice to see that you still care about the users of old Macs.
I did try to build OpenVPN 2.4 on 10.5, but without success so far.

Unfortunately the updater is broken again (probably a consequence of GitHub dropping TLS 1.0). I opened a new issue for that #442.
